商品購入機能

sorairo580 opened the pull request:
# WHAT
商品購入機能の実装

# WHY
商品購入機能実装のため
・必要な情報を適切に入力して「購入」ボタンを押すと、商品の購入ができる動画
https://gyazo.com/896043f1620c4f24810a1a9aa2114086
・入力に問題がある状態で「購入」ボタンが押された場合、情報は受け入れられず、購入ページでエラーメッセージが表示される動画
https://gyazo.com/ec90c769587789fb3e59190f486388b3
・ログイン状態の場合でも、URLを直接入力して自身が出品していない売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画
https://gyazo.com/d72564044b1bd572c4b42caab8fcd053
・ログイン状態の場合でも、URLを直接入力して自身が出品した商品の商品購入ページに遷移しようとすると、商品の販売状況に関わらずトップページに遷移する動画
https://gyazo.com/253873a95543ed8787db716c2100715e
・ログアウト状態の場合は、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移する動画
https://gyazo.com/877e54f46572d0ac6ed914dbc00554e7
・売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品一覧機能実装時に未実装であった場合）
https://gyazo.com/6dd21e7f7e5d249a67d5e0357b35ced4
・売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/862c40335d310facb96913018a36af15
・ログイン状態の場合でも、売却済みの商品には、「商品の編集」「削除」「購入画面に進む」ボタンが表示されない動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/0b8a483ca39bb1a531b8a9772e2fc11d
・ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）
https://gyazo.com/b49c30777e6924ce876bb796e5e08128
・テスト結果の画像
https://gyazo.com/67680bbe1ca7bb2100e6284dbcf172d6

sorairo580 committed:
purchase_recordモデル作成

sorairo580 committed:
delivery_destinationモデル作成

sorairo580 committed:
ルーティング設定

sorairo580 committed:
商品購入画面への遷移

purchase_recordに関すること以外はコメントアウト
ルーティングのネストを見直し

sorairo580 committed:
支払金額表示

sorairo580 committed:
purchase_recordモデルの編集

注文時のインスタンス名をorderへ変更
ルーティング見直し

sorairo580 committed:
PurchaseRecordsControllerへcreateアクション追加

sorairo580 committed:
DeliveryDestinationモデル編集

sorairo580 committed:
purchase_deliveryモデルなど作成

controllerも手動作成、過去を参考に記述
一つのフォームから２つ以上のテーブルへのデータ保存実施
インスタンス変数の定義方法などに注意

sorairo580 committed:
不要な記述のコメントアウト

sorairo580 committed:
購入機能モデル単体テスト１

sorairo580 committed:
テストコードクリア

sorairo580 committed:
JavaScript準備１

sorairo580 committed:
エラーハンドリング

sorairo580 committed:
JavaScript導入

購入画面のCSSのディレクトリ変更
pay.jp導入、v2仕様へindexを変更
tokenのテスト

sorairo580 committed:
payjpの設定

環境定数へ変更済み
未ログイン状態で購入ページへ直接遷移しないよう修正

sorairo580 committed:
payjpの実装とテスト修正

sorairo580 committed:
購入履歴と配送先の保存

sorairo580 committed:
Sold Out!の表示

sorairo580 committed:
購入済み商品の編集・購入ページへの遷移を制限

ShotaOtake commented on app/models/purchase_delivery.rb:
```diff
@@ -0,0 +1,32 @@
+class PurchaseDelivery
+  include ActiveModel::Model
+  attr_accessor :card_number, :card_exp, :card_cvc, :user_id, :item_id, :zip_cord,
```
◯不要な内容を削除しましょう

```
card_number, :card_exp, :card_cvc,
```

ShotaOtake commented on app/models/user.rb:
```diff
@@ -5,6 +5,7 @@ class User < ApplicationRecord
          :recoverable, :rememberable, :validatable
 
```
※別ファイルの指摘内容です。

商品モデルファイルにて購入モデルとのアソシエーションは記述いただいておりますでしょうか？
未設定の場合アソシエーションを付与しましょう。

記述済みの場合はご放念ください。

ShotaOtake commented on spec/factories/purchase_delivery.rb:
```diff
@@ -0,0 +1,17 @@
+    phone_number       { Faker::Number.between(from: 1111111111, to: 9999999999) }
+    token              {"tok_abcdefghijk00000000000000000"}
+
+    association :user, factory: :user, strategy: :build
```
◯アソシエーションを削除しましょう！

```
    association :user
    association :item
item_id
```

→フォームオブジェクト内でのアソシエーションの使用は推奨されないため

specファイルのbeforeブロックでインスタンス生成を行い、各情報へ値を持たせる実装に修正を行いましょう！

ShotaOtake commented on spec/models/purchase_delivery_spec.rb:
```diff
@@ -0,0 +1,96 @@
+require 'rails_helper'
```
■テストの要件を追加しましょう！

現状、下記の要件が満たされていない状況でございます。

```


・電話番号が空では購入できない
・電話番号が9桁以下では購入できない
・電話番号が12桁以上では購入できない
・電話番号に半角数字以外が含まれている場合は購入できない（※半角数字以外が一文字でも含まれていれば良い）

・userが紐付いていなければ購入できない

・itemが紐付いていなければ購入できない


※電話番号については上記のようにそれぞれ別のexampleとしてテストを行うように修正しましょう！

```

上記を全て満たせるようexampleを追加しましょう！

ShotaOtake reviewed (commented):
レビュー依頼を頂きありがとうございます！
追加で修正依頼のコメントをさせて頂きましたので、ご確認の上修正をお願い致します。

完成まであと一歩です！
引き続き実装頑張っていきましょう！

```
※テストコードにて修正箇所のございます際は、お手数ですが、最新の実行結果の画像も改めて添付頂きますようお願い致します。
```

sorairo580 committed:
プルリクエスト修正

・不要なコード削除
・spec/factories/purchase_delivery.rbのアソシエーション削除
・テストコード見なおし
・rubocop使用

sorairo580 commented:
修正を行いました。
ご確認をお願いします。
・テストコードの画像
https://gyazo.com/8d3f76c488215767f2c24dbf74842cec

yuri240 commented on app/models/purchase_delivery.rb:
```diff
@@ -0,0 +1,25 @@
+    validates :city
+    validates :house_number
+    validates :phone_number, format: { with: /\A\d{10,11}\z/, message: 'is invalid' }
+  end
```
`with_options presence: true do`という記述が2回存在しているため、
一回の記述で済むようにコードをまとめましょう！

理由：可読性向上のためです。

yuri240 commented on spec/models/purchase_delivery_spec.rb:
```diff
@@ -0,0 +1,96 @@
+      end
+      it '郵便番号が「３桁‐４桁」の文字列で入力されている' do
+        expect(@purchase_delivery).to be_valid
+      end
```
```ruby
      it 'クレジットカード情報（token）が存在し、配送先情報がすべて入力されている' do
        expect(@purchase_delivery).to be_valid
      end
      it '郵便番号が「３桁‐４桁」の文字列で入力されている' do
        expect(@purchase_delivery).to be_valid
      end
```
上記のテスト内容を拝見すると同じ意味を示しているため、
どちらか一方に修正しましょう！

yuri240 reviewed (commented):
レビュー依頼ありがとうございます！
キャプチャ画像と修正点確認させていただきました。

追加でコメントさせていただきましたので、ご確認お願いします。

sorairo580 committed:
プルリクエスト修正２

・purchase_delivery.rbの可読性意識
・同様のテスト内容の削除

sorairo580 commented:
修正2を行いました。
ご確認をお願いします。

ShotaOtake commented:
レビュー依頼を頂きありがとうございます！
お送り頂いた内容を確認致しました。

今回【LGTM】とさせて頂きます。
実装お疲れ様でした！

```
※現在のブランチはマージ後、ローカルのmasterブランチに忘れずにプルしましょう！
その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。
```

引き続き実装頑張っていきましょう！
応援しております！

sorairo580 commented:
ありがとうございました！
